src/loader/loader.ts

```diff
@@ -90,24 +128,25 @@ class ChibiLoader {
                         dispatch.addWorker(ExtensionWorker);
                     });
                 case 'unsandboxed': {
-                    const response = await fetch(ext);
-                    const originalScript = await response.text();
-                    const closureFunc = new Function('Scratch', originalScript);
                     const ctx = makeCtx(this.vm);
                     /*
                      * Provide compatibility support for some extensions
                      * Nevertheless, it is a very bad behavior for Chibi to directly read the Scratch
                      * object on the window, because the extension should not depend on unknown external
                      * environment.
                      */
-                    if (!('Scratch' in window) && !window.chibi.settings.dontExposeCtx) {
+                    const originalScratch = window.Scratch;
+                    if (!window.chibi.settings.dontExposeCtx) {
                         window.Scratch = ctx;
                     }
                     ctx.extensions.register = (extensionObj: ExtensionClass) => {
                         const extensionInfo = extensionObj.getInfo();
                         this._registerExtensionInfo(extensionObj, extensionInfo, ext);
                     };
-                    closureFunc(ctx);
+                    await loadChibiExtension(ext, ctx);
+                    if (!window.chibi.settings.dontExposeCtx) {
+                        window.Scratch = originalScratch;
```

Comment:
The ``Scratch`` object is not necessarily used only during the extension loading phase. In fact, it may be called by the extension at any time, so we should not restore the object.